engine/baml-lib/jsonish/src/deserializer/test_iterative_parser.rs

```diff
@@ -0,0 +1,279 @@
+test_jsonish!(test_trailing_comma_array, r#"[1, 2, 3,]"#, ["1", "2", "3"]);
+test_jsonish!(test_trailing_comma_object, r#"{"key": "value",}"#, {"key": "value"});
+
+// Test cases for invalid JSONish
+test_jsonish!(test_invalid_array, "[1, 2, 3", ["1", "2", "3"]);
+
+test_jsonish!(test_invalid_array_in_object, r#"{"key": [1, 2, 3"#, {"key": ["1", "2", "3"]});
+
+// Extra quote is not allowed
+test_jsonish!(test_incomplete_string, r#""hello"#, "hello");
+
+test_jsonish!(test_incomplete_string_in_object, r#"{"key": "value"#, {"key": "value"});
+
+// This is un-changed
+test_jsonish!(
+    test_prefixed_incompleted_string,
+    r#"prefix "hello"#,
+    r#"prefix "hello"#
+);
+
+test_jsonish!(
+  test_large_object,
+  r#"{"key": "value", "array": [1, 2, 3], "object": {"key": "value"}}"#,
+  {
+    "key": "value",
+    "array": [1, 2, 3],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_md_example,
+  r#"
+  some text
+  ```json
+  {
+    "key": "value",
+    "array": [1, 2, 3],
+    "object": {
+      "key": "value"
+    }
+  }
+  ```
+  "#,
+  {
+    "key": "value",
+    "array": [1, 2, 3],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_md_example_2,
+  r#"
+  some text
+  ```json
+  {
+    "key": "value",
+    "array": [1, 2, 3],
+    "object": {
+      "key": "value"
+    }
+  }
+  ```
+
+  ```json
+  ["1", "2"]
+  ```
+  "#,
+  [{
+    "key": "value",
+    "array": [1, 2, 3],
+    "object": {
+      "key": "value"
+    }
+  }, ["1", "2"]]
+);
+
+test_jsonish!(
+  test_json_md_bad_example,
+  r#"
+  some text
+  ```json
+  {
+    "key": "value",
+    "array": [1, 2, 3,],
+    "object": {
+      "key": "value"
+    },
+  }
+  ```
+  "#,
+  {
+    "key": "value",
+    "array": ["1", "2", "3"],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_md_bad_example_2,
+  r#"
+  some text
+  ```json
+  {
+    "key": "value",
+    "array": [1, 2, 3, "some stinrg"with quotes"],
+    "object": {
+      "key": "value"
+    },
+  }
+  ```
+  "#,
+  {
+    "key": "value",
+    "array": ["1", "2", "3", "some stinrg\"with quotes"],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_md_bad_example_single_quote,
+  r#"
+  some text
+  ```json
+  {
+    "key": "value",
+    "array": [1, 2, 3, 'some stinrg'   with quotes' /* test */],
+    "object": { // Test comment
+      "key": "value"
+    },
+  }
+  ```
+  "#,
+  {
+    "key": "value",
+    "array": ["1", "2", "3", "some stinrg'   with quotes"],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_with_unquoted_keys,
+  r#"
+  {
+    key: "value",
+    array: [1, 2, 3],
+    object: {
+      key: "value"
+    }
+  }
+  "#,
+  {
+    "key": "value",
+    "array": ["1", "2", "3"],
+    "object": {
+      "key": "value"
+    }
+  }
+);
+
+test_jsonish!(
+  test_json_with_unquoted_values_with_spaces,
```

Comment:
@villagab4 this passes btw

if the string fix works for you, i'll pause on this for now to mitigate errors